src/app/plugins/commandsearch.plugin.ts

```diff
@@ -0,0 +1,61 @@
+import { Plugin } from '../../common/plugin';
+import { ChannelType, IContainer, IMessage, IPlugin } from '../../common/types';
+
+export class CommandSearchPlugin extends Plugin {
+  public name: string = 'Command Search';
+  public description: string = 'Search our commands on a key word or phrase';
+  public usage: string = 'commands <word or phrase>';
+  public pluginAlias = ['searchcommands', 'grep'];
+  public permission: ChannelType = ChannelType.Bot;
+  public commandPattern: RegExp = /[^]+/;
+
+  constructor(public container: IContainer) {
+    super();
+  }
+
+  public async execute(message: IMessage, args: string[]) {
+    const query = args.join(' ');
+
+    // For every plugin, evaluate it's match
+    const results = Object.entries(this.container.pluginService.plugins).reduce(
+      (results: string[], [pluginName, plugin]) => {
+        if (
+          plugin.permission === ChannelType.Admin ||
+          plugin.permission === ChannelType.Staff ||
+          !this._grep(plugin, query)
+        ) {
+          return results;
+        }
+
+        results.push(pluginName);
+        return results;
+      },
+      []
+    );
+
+    if (!results.length) {
+      message.reply(`I couldn't find any results for that query.`);
+      return;
+    }
+
+    const embeds = this.container.pluginService.generateHelpEmbeds(results, 'adv');
+    embeds.forEach((embed) =>
+      embed.setTitle('**__I found the following commands matching your search__**')
+    );
+    this.container.messageService.sendPagedEmbed(message, embeds);
+  }
+
+  private _grep(plugin: IPlugin, query: string): boolean {
+    const pluginMeta = `
+      ${plugin.name}
+      ${plugin.description}
+      ${plugin.usage}
+      ${plugin.pluginAlias?.join(' ')}`.toLowerCase();
+
+    // Find greplike matches within plugin's metadata.
+    const grepRegex = new RegExp(`^.*(${query.toLowerCase()}).*$`, 'mg');
+    const match = pluginMeta.match(grepRegex);
```

Comment:
no need to build a regex for this search

Reply:
I think the idea was that a user could make their own regex as input too
What would be nice, is if when you build the pluginMeta, you did something like ` \`name: ${plugin.name}\`` and so on,
that way each element is query-able with regex @TylerMathis

Reply:
It would be cool to support custom regex, why not? :D